### ISSUE TITLE
closing labwc subsession with either wayland or X11 backend doesn't end labwc process

### ISSUE THREAD

**Issue by jonleivent** [labwc/labwc#946]
Using labwc 0.6.3 on debian 12.

When closing a labwc subsession window started with either a wayland or X11 session backend, the subsession labwc window vanishes but the labwc process does not exit.

BTW: I suspect it's the case that running any wayland compositor as a subsession isn't commonly done.  I've seen more problems with subsessions for other compositors I've tried (weston, cage, kwin) than occur for normal console-started sessions.  Most of those occur when resizing the subsession window.  Labwc doesn't seem to have those resizing problems.

**Comment by Consolatis** [comment]
labwc / wlroots are completely happy to run with no outputs connected. You wouldn't want labwc to terminate just because you physically disconnected your monitor either I assume. But I agree that it feels a bit weird to have to `Ctrl-C` the nested instance after closing the output window. I guess we could test if we are currently running nested (e.g. using the wlroots X11 / wayland backend) and if all outputs are gone: terminate.

It will get a bit trickier once we end up implementing "dummy" outputs for use from remote VNC connections as there will then always be some output left.

**Comment by jonleivent** [comment]
Would make sense instead to add an `--exit-with-last-client` option to labwc, similar to what X servers have with their `-terminate` option?

**Comment by Consolatis** [comment]
I guess that could be done. We'd need to define what counts as client though, e.g. is a panel a "client" regarding the option?
Same for `swaybg` and friends.

**Comment by jonleivent** [comment]
> What is your actual use-case if I may ask? You are running specific applications within their own nested compositor to create security boundaries?

I am using labwc as a desktop compositor, but I am also looking for a way to use a nested compositor for certain apps running from other user accounts or for apps running in sandboxes.  These would be single apps per compositor.  It's not clear to me, however, that I can accomplish this fully even if I can get a nested compositor to work properly.  In the sandboxing case, I'd like to completely isolate the sandboxed app from access to the external /dev/shm and /run/user (or any other way wayland compositors like to share info with clients).  I want the clipboards to be separate as well.

**Comment by Consolatis** [comment]
Maybe [Cage](https://github.com/cage-kiosk/cage) would be a better fit as the nested compositor?

Regarding `/dev/shm` and friends: it sounds like you could make some use of mount namespaces, that would (depending on kernel settings and system setup, e.g. `/etc/sub{g,u}id` ) also work for unprivileged user accounts by
- first creating their own uid/guid namespace so they become pseudo root
- using that pseudo root to create a new mount namespace
- dropping pseudo root by becoming some unprivileged user within that uid/gid namespace
- finally running the application

Doing that manually involves a bit of work and reading but I think there are also sandbox engines which can simplify the setup.
In either case I think the main point here is to start cage in the main mount namespace so it has access to the labwc wayland socket and then bind-mounting the wayland socket of the nested compositor (e.g. cage) into the unprivileged mount namespace. That way an application running within the nested compositor shouldn't be able to access anything of value anymore (unless you bind-mounted it into its own mount namespace while setting it up).

**Comment by johanmalm** [comment]
In response to the OP, I think it depends on how the nested compositor is exited. When choosing exit from the root-menu it exits without having to ctrl+C in the parent terminal.

**Comment by jonleivent** [comment]
> Maybe [Cage](https://github.com/cage-kiosk/cage) would be a better fit for the nested compositor?

As menitioned, I tried cage.  It has bugs when resizing as a nested compositor.   I will post an issue there about that...

> Regarding /dev/shm and friends: it sounds like you could make some use of mount namespaces...

I do all of that already, using bubblewrap.  The issue is that with wayland compositors and certain clients, /dev/shm and friends have to be shared so that the client can create something there that can be shared with the compositor.  This means that the /dev/shm that the compositor sees must be mounted in the sandbox's mount namespace and writable by the client.  I was hoping that by having a separate nested compositor,  that the backend compositor (or X11) would not need to share /dev/shm with the client of the nested compositor.  But that would only work if the nested compositor didn't need to share /dev/shm with its backend.

In other words, for wayland clients that insist on creating shared memory files and sharing them with their compositor, there's no way to sandbox clients of the same compositor from each other.  All clients of the same compositor must share the same directory (usually /dev/shm), and so all can see (and hack) each other's stuff.  For clients that don't insist on using shared memory, there's no problem: share only the wayland socket into the sandbox.  I don't need a nested compositor for those.

Maybe waypipe is needed instead.


> In response to the OP, I think it depends on how the nested compositor is exited. When choosing exit from the root-menu it exits without having to ctrl+C in the parent terminal.

That's correct.  The case I had tested is when exiting the nested compositor by killing its window.

**Comment by Consolatis** [comment]
> Would make sense instead to add an `--exit-with-last-client` option to labwc, similar to what X servers have with their `-terminate` option?

I don't think that we should implement this, its too difficult to define what an `client` actually is as written above (swaybg / notification daemon / some panel / ..).

> I guess we could test if we are currently running nested (e.g. using the wlroots X11 / wayland backend) and if all outputs are gone: terminate.

That we could actually do. And regarding headless outputs we could simply say that we don't exit as long as an output exists (including headless ones). So if people want labwc to exit when closing the last output windows they should simply not set up headless outputs in the first place (which aren't support currently anyway).

**Comment by dyfrgi** [comment]
I have a use-case for this feature - the original feature discussed here where labwc terminates when its last output is closed, not the one related to client windows in labwc.

I am using labwc to run a game and an overlay for that game (Path of Exile and Awakened PoE Trade). I need these to be in the same rootful Xwayland because the windows must be positioned relative to each other, hotkeys need to be bindable by both programs, and the overlay needs to be able to send keyboard input to the game. These are capabilities that are simple in X and only possible on a very few Wayland compositors otherwise, with some custom configuration for the setup.

I could probably run XWayland + an X11 window manager, which might give better latency, but so far nested labwc is working. With labwc, I start labwc, then the game, then the overlay, and things (mostly) Just Work, barring some weird offsets in mouse locations received by the overlay which thankfully can be easily worked around - I can still click on things, it's just not good at knowing when one mouseover event happens.

What I would like to be able to do is to close the nested compositor and have all the clients close as well. Presently, when I close the window, labwc just logs a message but the clients don't exit since they're still connected to something. If labwc instead terminated when its output window was closed, this would work.

**Comment by ih8ms** [comment]
I'm starting Steam with labwc autostart. My use case was to exit labwc when I exit steam (instead of ending up with a blank screen). My solution was simply this in autostart:

`steam ; pkill labwc`

Seems to work reasonable enough.

**Comment by Consolatis** [comment]
You could also use `labwc --session steam` instead. That has the benefit that it doesn't accidentally kill other labwc instances.

**Comment by Consolatis** [comment]
Implemented via
- #3440
